packages/react-reconciler/src/SchedulingProfiling.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Lane, Lanes} from './ReactFiberLane';
+import type {Fiber} from './ReactInternalTypes';
+
+import {enableSchedulingProfiling} from 'shared/ReactFeatureFlags';
+import {getStackByFiberInDevAndProd} from './ReactFiberComponentStack';
+
+/**
+ * If performance exists and supports the subset of the User Timing API that we
+ * require.
+ */
+const supportsUserTiming =
+  typeof performance !== 'undefined' && typeof performance.mark === 'function';
+
+function formatLanes(laneOrLanes: Lane | Lanes): string {
+  return ((laneOrLanes: any): number).toString();
+}
+
+export function markCommitStarted(lanes: Lanes): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-start-${formatLanes(lanes)}`);
+    }
+  }
+}
+
+export function markCommitStopped(): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-stop`);
+    }
+  }
+}
+
+const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
+// $FlowFixMe: Flow cannot handle polymorphic WeakMaps
+const wakeableIDs: WeakMap<Wakeable, number> = new PossiblyWeakMap();
+let wakeableID: number = 0;
+function getWakeableID(wakeable: Wakeable): number {
+  if (!wakeableIDs.has(wakeable)) {
+    wakeableIDs.set(wakeable, wakeableID++);
+  }
+  return ((wakeableIDs.get(wakeable): any): number);
+}
+
+export function markComponentSuspended(
+  componentName: string,
+  fiber: Fiber,
+  wakeable: Wakeable,
+): void {
+  if (enableSchedulingProfiling) {
+    if (supportsUserTiming) {
+      const id = getWakeableID(wakeable);
+      const componentStack = getStackByFiberInDevAndProd(fiber) || '';
+      // TODO (brian) Generate and store temporary ID so DevTools can match up a component stack later.
+      performance.mark(
+        `--suspense-suspend-${componentName}-${id}-${componentStack}`,
+      );
+      wakeable.then(
+        () =>
+          performance.mark(
+            `--suspense-resolved-${componentName}-${id}-${componentStack}`,
```

Comment:
Do we have plans to use these ids in the profiler/visualizations? To be honest, I haven't thought about trying to dedup promises in the profiler UI. Maybe we don't need to maintain this map for now? At least not until we have a concrete idea for how we'd use the data?

Reply:
I haven't thought about it either, but I think it could be useful if we highlighted all suspense events related to the wakeable when one of them is hovered over? It's currently not possible to tell when a suspense event resolves/rejects, especially if there are many instances of the same component suspending in quick succession, which I'd imagine will be a very common case.

![image](https://user-images.githubusercontent.com/12784593/86306209-3ddf5300-bc46-11ea-808e-043a123c869c.gif)

Reply:
Okay. If we plan to use the data (and the example you mention seems reasonable) then 👍